mlir/test/fusion/linalg-generic-reuse-buffers.mlir

```diff
@@ -0,0 +1,60 @@
+    %alloc = memref.alloc() : memref<64x64x64xf32>
+    linalg.generic {indexing_maps = [#map1, #map1, #map1], iterator_types = ["parallel", "parallel", "parallel"]} ins(%arg0, %arg3 : memref<64x64x64xf16>, memref<64x64x64xf32>) outs(%alloc : memref<64x64x64xf32>) {
+    ^bb0(%in: f16, %in_10: f32, %out: f32):
+      %46 = arith.extf %in : f16 to f32
+      %47 = arith.addf %46, %in_10 : f32
+      linalg.yield %47 : f32
+    }
+    %c0 = arith.constant 0 : index
+    %1 = rock.workgroup_id : index
+    %2 = rock.workitem_id : index
+    %3 = rock.transform %alloc by #transform_map : memref<64x64x64xf32> to memref<4x64x1x1x16x16x4x1xf32>
+    %4 = rock.transform %3 by #transform_map1 : memref<4x64x1x1x16x16x4x1xf32> to memref<4x64x1x1x256x4xf32>
+    %out = rock.alloc() : memref<4xf32, #gpu.address_space<private>>
+    %c1_1 = arith.constant 1 : index
+    %5 = arith.divui %1, %c1_1 : index
+    // CHECK: [[tiled_arg0:%.+]] = rock.alloc() : memref<4xf16
+    // CHECK-NEXT: [[arg0_tr:%.+]] = rock.transform %arg0 by [[MAP]]
+    // CHECK-NEXT: [[arg0_view:%.+]] = rock.transform [[arg0_tr]] by [[MAP1]]
+    // CHECK-NEXT: rock.threadwise_read_into {{.*}} []([[arg0_view]]) [{{.*}}] -> [[tiled_arg0]]
+    // CHECK: [[tiled_arg3:%.+]] = rock.alloc() : memref<4xf32
+    // CHECK-NEXT: [[arg3_tr:%.+]] = rock.transform %arg3 by [[MAP]]
+    // CHECK-NEXT: [[arg3_view:%.+]] = rock.transform [[arg3_tr]] by [[MAP1]]
+    // CHECK-NEXT: rock.threadwise_read_into {{.*}} []([[arg3_view]]) [{{.*}}] -> [[tiled_arg3]]
+    // CHECK: [[tiled_out:%.+]] = rock.alloc() : memref<4xf32
+    // CHECK-NEXT: linalg.generic
+    // CHECK-SAME: ins([[tiled_arg0]], [[tiled_arg3]]
+    // CHECK-SAME: outs([[tiled_out]]
+    rock.threadwise_read_into {forceUnroll, useIndexDiffs} [](%4) [%c0, %2, %2, %5, %5] -> %out : memref<4x64x1x1x256x4xf32> -> memref<4xf32, #gpu.address_space<private>>
+    // CHECK: [[arg0_alloc:%.+]] = rock.alloc() : memref<4xf16
+    // CHECK: [[arg3_alloc:%.+]] = rock.alloc() : memref<4xf32
+    affine.for %arg4 = 1 to 4 {
+      // CHECK: [[arg0_tf:%.+]] = rock.transform %arg0 by [[MAP]]
+      // CHECK-NEXT: [[arg0_v:%.+]] = rock.transform [[arg0_tf]] by [[MAP1]]
+      // CHECK-NEXT: rock.threadwise_read_into {{.*}} []([[arg0_v]]) [%arg4, {{.*}}] -> [[arg0_alloc]]
+      // CHECK: [[arg3_tf:%.+]] = rock.transform %arg3 by [[MAP]]
+      // CHECK-NEXT: [[arg3_v:%.+]] = rock.transform [[arg3_tf]] by [[MAP1]]
+      // CHECK-NEXT: rock.threadwise_read_into {{.*}} []([[arg3_v]]) [%arg4, {{.*}}] -> [[arg3_alloc]]
+      // CHECK: linalg.generic {indexing_maps = [#map2, #map2, #map2], iterator_types = ["parallel"]} ins(%12, %13 : memref<4xf16, #gpu.address_space<private>>, memref<4xf32, #gpu.address_space<private>>) outs(%11 : memref<4xf32, #gpu.address_space<private>>) {
```

Comment:
Ok, so, to ask very annoying questions: ... how is it your code ensures that the same temporary allocation gets used for both the generic in the loop and the generic outside of it? I didn't see anything in your changes that would've caused the two `rock::AllocOp`s to merge together.

One possibility I'm concerned by is that you're not calling `clone` early enough, and thus assuming that all the users of a given `linalg.generic` are tiled the same.

For example, let's do another test where we add a second `threadwise_read_into` that has a different tile size than the one outside the loop.

Reply:
> Ok, so, to ask very annoying questions: ... how is it your code ensures that the same temporary allocation gets used for both the generic in the loop and the generic outside of it? I didn't see anything in your changes that would've caused the two `rock::AllocOp`s to merge together.
> 
I didn't merge them. The two threadwise_read_into in the original IR happen to write into the same allocation, so the two generic use the same allocation as well. If the original threadwise_read_into ops write into different allocations, the generic ops would use different temporary allocations, even if they have the same tile size.